pkg/util/util.go

```diff
@@ -0,0 +1,346 @@
+	if err != nil {
+		glog.Fatalf("Failed to create ObjectStorage Controller: %v", err)
+	}
+
+	go ctrl.Run(ctx)
+	return client, kubeClient, ctrl
+}
+
+func GetBuckets(ctx context.Context, client bucketclientset.Interface, numExpected int) *types.BucketList {
+	bucketList, err := client.ObjectstorageV1alpha1().Buckets().List(ctx, metav1.ListOptions{})
+	if len(bucketList.Items) > 0 {
+		return bucketList
+	}
+	numtimes := 0
+	for numtimes < 10 {
+		bucketList, err = client.ObjectstorageV1alpha1().Buckets().List(ctx, metav1.ListOptions{})
+		if len(bucketList.Items) >= numExpected {
+			return bucketList
+		} else {
+			glog.Errorf("Failed to fetch the bucket created %v", err)
+		}
+		numtimes++
+		<-time.After(time.Duration(numtimes) * time.Second)
+	}
+	return &types.BucketList{}
+}
+
+func ValidateBucket(bucket types.Bucket, bucketrequest types.BucketRequest, bucketclass types.BucketClass) bool {
+	if strings.HasPrefix(bucket.Name, bucketrequest.Spec.BucketPrefix) &&
+		bucket.Spec.BucketClassName == bucketrequest.Spec.BucketClassName &&
+		bucket.Spec.BucketRequest.Name == bucketrequest.Name &&
+		bucket.Spec.BucketRequest.Namespace == bucketrequest.Namespace &&
+		bucket.Spec.BucketRequest.UID == bucketrequest.ObjectMeta.UID &&
+		bucket.Spec.BucketClassName == bucketclass.Name &&
+		reflect.DeepEqual(bucket.Spec.Parameters, bucketclass.Parameters) &&
+		bucket.Spec.Provisioner == bucketclass.Provisioner &&
+		bucket.Spec.RetentionPolicy == bucketclass.RetentionPolicy &&
+		bucket.Spec.AnonymousAccessMode == bucketclass.AnonymousAccessMode {
+		return true
+	}
+	return false
+}
+
+func GetBucketAccesses(ctx context.Context, client bucketclientset.Interface, numExpected int) *types.BucketAccessList {
+	bucketaccessList, _ := client.ObjectstorageV1alpha1().BucketAccesses().List(ctx, metav1.ListOptions{})
+	if len(bucketaccessList.Items) > 0 {
+		return bucketaccessList
+	}
+	numtimes := 0
+	for numtimes < 10 {
+		bucketaccessList, _ = client.ObjectstorageV1alpha1().BucketAccesses().List(ctx, metav1.ListOptions{})
+		if len(bucketaccessList.Items) >= numExpected {
+			return bucketaccessList
+		}
+		numtimes++
+		<-time.After(time.Duration(numtimes) * time.Second)
+	}
+	return &types.BucketAccessList{}
+}
+
+func ValidateBucketAccess(bucketaccess types.BucketAccess, bucketaccessrequest types.BucketAccessRequest, bucketaccessclass types.BucketAccessClass) bool {
+	if bucketaccess.Spec.BucketInstanceName != "" &&
+		bucketaccess.Spec.BucketAccessRequest == bucketaccessrequest.Name &&
+		bucketaccess.Spec.ServiceAccount == bucketaccessrequest.Spec.ServiceAccountName &&
+		bucketaccess.Spec.PolicyActionsConfigMapData != "" &&
+		bucketaccess.Spec.Provisioner == bucketaccessclass.Provisioner {
+		return true
+	}
+	return false
+}
+
+func DeleteObjects(ctx context.Context, client bucketclientset.Interface, objs ...interface{}) {
```

Comment:
I'd keep some more specific `defer`s in the test code rather than a catch-all like this. Seems a bit error-prone.

Reply:
I believe this way code is less verbose in the test and test is more readable. I agree user need to remember all objects to delete whether he choses to call this multiple time with multiple defer calls or at once.

Reply:
It's 'magic'. Go code should be explicit and simple, no 'magic'. But fair enough, matter of taste so no real objection.